### DISCUSSION TITLE
[@types/vuelidate] incorrect type definition for email validator

### DISCUSSION THREAD

**Discussion by philipandresen** [DefinitelyTyped/DefinitelyTyped#61618] [Issues with a @types package]
I ran into the following issue while adding an email validator to a vue project. The email validator type definition is 
[/vuelidate/lib/validators.d.ts:50](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f7ec78508c6797e42f87a4390735bc2c650a1bfd/types/vuelidate/lib/validators.d.ts)
`export function email(): ValidationRule`

However the plain JS export in Vuelidate for the email function is as follows:
[vuelidate/packages/validators/src/raw/email.js:5](https://github.com/vuelidate/vuelidate/blob/next/packages/validators/src/raw/email.js)
`export default regex(emailRegex)`

the function `regex()` returns a `ValidationRule`, and since it is invoked in the export here, the default export for the email validator is actually a `ValidationRule` instead of a `(): ValidationRule`

I could instead suggest a fix in Vuelidate if people feel like that is more appropriate. If it were to export `() => regex(emailRegex)` the type def would fit nicely and all the exported validators would have the same shape. It's a little ambiguous if it's actually an issue with the library itself but it is *definitely* an issue with the typedef until that is changed in the underlying library.

For now, it's easy enough to just ignore the typescript warnings about the incorrect type, so this is not incredibly urgent. I also did a cursory glance at a random sampling of other validation functions and the others seem to be defined correctly. This seems to be an anomaly with just email.

[a comment by an automated account is not shown]

**Reply by janesser** [reply]
@philipandresen  i have been checking, as you see here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f7ec78508c6797e42f87a4390735bc2c650a1bfd/types/vuelidate/vuelidate-tests.ts#L82

`decimal` works same as `email`. dtslint nor vue-playground shows any errors. do you have a sandbox to reproduce your issue?

**Reply by philipandresen** [reply]
@janesser Thanks for taking a look. This might be crazy but I have never actually created a code sandbox before, so I hope this works for you:
https://codesandbox.io/s/confident-leavitt-vkxz5k?file=/src/index.ts 

Should see the following (also copying here for posterity)
```
const emailValidator: ValidationRule = email();
const decimalValidator: ValidationRule = decimal();
const regexValidator: ValidationRule = helpers.regex("PostalCode",/^[0-9]{5}(?:-[0-9]{4})?$/);
const requiredValidator: ValidationRule = required();
const minLengthValidator: ValidationRule = minLength(10);
const maxLengthValidator: ValidationRule = maxLength(200);

console.log(`${typeof emailValidator} Should be "function"`);     // boolean Should be "function"
console.log(`${typeof decimalValidator} Should be "function"`);   // boolean Should be "function"
console.log(`${typeof regexValidator} Should be "function"`);     // function Should be "function"
console.log(`${typeof requiredValidator} Should be "function"`);  // boolean Should be "function"
console.log(`${typeof minLengthValidator} Should be "function"`); // function Should be "function"
console.log(`${typeof maxLengthValidator} Should be "function"`); // function Should be "function"
```

I guess looking at this in more depth, the type definitions also appear to be wrong for all of the validator functions that take no arguments:
```
required()
alpha()
alphaNum()
numeric()
integer()
decimal()
email()
ipAddress()
url()
```
These should just be `ValidationRule`s instead of `() => ValidationRule`s

**Reply by philipandresen** [reply]
@janesser If you would like, I can open a PR with the fix.

**Reply by janesser** [reply]
I'm still not getting what is the difference between realworld testcase and the sandbox type reflexion. Please don't hesitate to propose an improvement. Notably some of the typings resolve fine.

**Reply by philipandresen** [reply]
> Notably some of the typings resolve fine.

@janesser Conversely, some of the typings **do not** resolve fine. Therein lies the issue. If I were to call `emailValidator()` in that sandbox, I would not get any typescript warnings but at runtime I would get `emailValidator is not a function` because it is NOT a function, it is a boolean. Typescript thinks it's a function though.

The sandbox demonstrates that despite all of those `const`s having the exact same type definition, at runtime they have very different types. This indicates that the type definitions are incorrect for

```
required()
alpha()
alphaNum()
numeric()
integer()
decimal()
email()
ipAddress()
url()
```
And, as you observed, the type definitions are correct for the others.
